File: 01-programming-languages/C#/02-oop/01-classes/classes.cs
// C# Classes

using System;

namespace ClassesDemo
{
    // Basic class
    public class Person
    {
        // Properties
        public string Name { get; set; }
        public int Age { get; set; }
        
        // Constructor
        public Person(string name, int age)
        {
            Name = name;
            Age = age;
        }
        
        // Method
        public string GetDetails()
        {
            return $"{Name}, Age: {Age}";
        }
        
        // Override ToString
        public override string ToString()
        {
            return $"Person: {Name}";
        }
    }
    
    // Class with encapsulation
    public class BankAccount
    {
        // Private field
        private decimal balance;
        
        // Public property with validation
        public decimal Balance
        {
            get { return balance; }
        }
        
        // Constructor
        public BankAccount(decimal initialBalance)
        {
            if (initialBalance < 0)
                throw new ArgumentException("Initial balance cannot be negative");
            balance = initialBalance;
        }
        
        // Methods
        public void Deposit(decimal amount)
        {
            if (amount <= 0)
                throw new ArgumentException("Deposit amount must be positive");
            balance += amount;
        }
        
        public void Withdraw(decimal amount)
        {
            if (amount <= 0)
                throw new ArgumentException("Withdrawal amount must be positive");
            if (amount > balance)
                throw new InvalidOperationException("Insufficient funds");
            balance -= amount;
        }
    }
    
    // Static class
    public static class MathHelper
    {
        public static double PI = 3.14159;
        
        public static double Add(double a, double b)
        {
            return a + b;
        }
        
        public static double Multiply(double a, double b)
        {
            return a * b;
        }
    }
    
    // Sealed class
    public sealed class Logger
    {
        private static Logger instance;
        
        private Logger() { }
        
        public static Logger Instance
        {
            get
            {
                if (instance == null)
                    instance = new Logger();
                return instance;
            }
        }
        
        public void Log(string message)
        {
            Console.WriteLine($"[LOG] {message}");
        }
    }
    
    // Abstract class
    public abstract class Shape
    {
        public string Name { get; set; }
        
        protected Shape(string name)
        {
            Name = name;
        }
        
        public abstract double Area();
        public abstract double Perimeter();
        
        public string Describe()
        {
            return $"{Name}: Area = {Area():F2}, Perimeter = {Perimeter():F2}";
        }
    }
    
    class Program
    {
        static void Main(string[] args)
        {
            // Basic class
            Person person = new Person("Alice", 30);
            Console.WriteLine(person.GetDetails());
            Console.WriteLine(person.ToString());
            
            // Encapsulation
            BankAccount account = new BankAccount(1000);
            account.Deposit(500);
            Console.WriteLine($"Balance: {account.Balance}");
            
            // Static class
            Console.WriteLine($"PI: {MathHelper.PI}");
            Console.WriteLine($"Add: {MathHelper.Add(2, 3)}");
            
            // Sealed class (singleton)
            Logger logger = Logger.Instance;
            logger.Log("Application started");
            
            Console.WriteLine("Classes example running");
        }
    }
}

File: 01-programming-languages/Kotlin/02-oop/01-classes/classes.kt
// Kotlin Classes - Primary constructor, init block

// Basic class
class Person(val name: String, val age: Int) {
    init {
        println("Person created: $name, $age")
    }
    
    fun greet(): String {
        return "Hello, I'm $name"
    }
}

// Class with secondary constructor
class User(val name: String, val email: String) {
    var age: Int = 0
        private set
    
    constructor(name: String, email: String, age: Int) : this(name, email) {
        this.age = age
    }
    
    override fun toString(): String {
        return "User(name=$name, email=$email, age=$age)"
    }
}

// Data class (auto-generates equals, hashCode, toString, copy)
data class Point(val x: Int, val y: Int)

// Sealed class
sealed class Result {
    data class Success(val data: String) : Result()
    data class Error(val message: String) : Result()
    object Loading : Result()
}

// Abstract class
abstract class Shape {
    abstract fun area(): Double
    abstract fun perimeter(): Double
    
    fun describe(): String {
        return "Area: ${area()}, Perimeter: ${perimeter()}"
    }
}

class Circle(val radius: Double) : Shape() {
    override fun area(): Double = Math.PI * radius * radius
    override fun perimeter(): Double = 2 * Math.PI * radius
}

class Rectangle(val width: Double, val height: Double) : Shape() {
    override fun area(): Double = width * height
    override fun perimeter(): Double = 2 * (width + height)
}

// Companion object (static members)
class MathUtils {
    companion object {
        const val PI = 3.14159
        
        fun add(a: Int, b: Int): Int = a + b
        fun multiply(a: Int, b: Int): Int = a * b
    }
}

// Object declaration (singleton)
object Database {
    private val connections = mutableListOf<String>()
    
    fun connect(url: String) {
        connections.add(url)
        println("Connected to $url")
    }
    
    fun disconnect(url: String) {
        connections.remove(url)
        println("Disconnected from $url")
    }
    
    fun getConnectionCount(): Int = connections.size
}

fun main() {
    // Basic class
    val person = Person("Alice", 30)
    println(person.greet())
    
    // Secondary constructor
    val user = User("Bob", "bob@example.com", 25)
    println(user)
    
    // Data class
    val point1 = Point(1, 2)
    val point2 = Point(1, 2)
    println("Points equal: ${point1 == point2}")
    println("Point: $point1")
    
    // Copy function
    val point3 = point1.copy(x = 10)
    println("Copied: $point3")
    
    // Destructuring
    val (x, y) = point1
    println("Destructured: x=$x, y=$y")
    
    // Sealed class
    val result: Result = Result.Success("Data loaded")
    when (result) {
        is Result.Success -> println("Success: ${result.data}")
        is Result.Error -> println("Error: ${result.message}")
        is Result.Loading -> println("Loading...")
    }
    
    // Abstract class
    val circle = Circle(5.0)
    println(circle.describe())
    
    val rect = Rectangle(4.0, 6.0)
    println(rect.describe())
    
    // Companion object
    println("PI: ${MathUtils.PI}")
    println("Add: ${MathUtils.add(2, 3)}")
    
    // Singleton
    Database.connect("jdbc:mysql://localhost/db")
    Database.connect("jdbc:postgresql://localhost/db")
    println("Connections: ${Database.getConnectionCount()}")
    
    println("Classes example running")
}